### DISCUSSION TITLE
Best way to implement audio streaming?

### DISCUSSION THREAD

**Discussion by Azorlogh** [pdeljanov/Symphonia#153] [Q&A] [answered]
Hey there, I'm making a program that needs to audio streams from the internet, and I've hit some challenges using Symphonia with HTTP readers, because I cannot block on the audio thread but the reader might not have any data available.

My general strategy until now was to pre-fill a buffer with some data on a separate thread, and hook Symphonia to that buffer.
If the data is not available, I return a WouldBlock, which gets passed through by Symphonia and gets interpreted by my audio routine to return silence until some time has passed to try again.

This works for normal audio streaming, but requires additional logic when first creating the source (ensuring enough data is available for the FormatReader to initialize), and when seeking (making sure the data at the next location is available).

For some formats this is easy: `Mp3Reader` only needs a few bytes prepared in the buffer to begin working, and for seeking, I can just pre-fetch a bunch of bytes at an arbitrary location before I tell Symphonia to seek there. (Well, Mp3Reader will need SeekMode::Coarse for that but until then I'm just restarting Symphonia there)

However for more complex formats, like `IsoMp4Reader`, this strategy gets tricky. Upon creation, it tries to fetch data from all around the stream and there's no way for me to know what to load in advance to avoid blocking. That is not too bad, I can always make the source creation blocking since it doesn't happen on the audio thread.
But for seeking, things are more subtle. Again I don't know which bytes will need to be read, so I can't pre-fill the buffer and need to make seeking blocking. But that means I have to move my source to a separate thread to do this operation then move it back to the audio thread which seems clunky.

Another strategy I thought of is to just put Symphonia in a separate thread from the audio buffer so that it can block. But I'm not sure how wasteful it would be to transfer the audio samples across threads, instead of only compressed data, so I'm hesitant.

Does any of these approaches make sense or is there an easier possibility that I'm missing?

I think if Symphonia supported resuming after WouldBlock errors, it would make things easier for me (was also mentioned here: https://github.com/pdeljanov/Symphonia/issues/27#issuecomment-1191999584) because then I wouldn't need complex pre-fetching and could just return silence when I see a WouldBlock. But I assume that would require quite a bit more logic on Symphonia's side to handle this.

**Comment by jpochyla** [comment, marked as answer]
Hi @Azorlogh, I also encountered this problem when integrating Symphonia into [Psst](https://github.com/jpochyla/psst). I think it works quite well, but I have zero experience with audio processing, so beware (@pdeljanov would be the authoritative voice here). I also might have over-engineered things a little.

Related sources:
 - Audio output, generic audio source: https://github.com/jpochyla/psst/tree/master/psst-core/src/audio. Output reads from audio source in a dedicated thread, checks a channel in the audio callback for atomic switches to a new source.
 - Worker: https://github.com/jpochyla/psst/blob/master/psst-core/src/player/worker.rs. Worker is an actor (just a thread with some communication boilerplate removed), reads the decoded data from Symphonia and pushes them onto a ring-buffer). Supports seeking and reporting the play-head position.
 - Storage: https://github.com/jpochyla/psst/blob/master/psst-core/src/player/storage.rs Keeps track of what is already downloaded and what isn't, basically.

**Reply by Azorlogh** [reply]
Wow thanks a lot for the swift response. That's an amazing project you've got there! I'm going to take a close look at your architecture.

For the record, what I'm making is a music client meant to accept multiple sources (mainly soundcloud, youtube, spotify). It doesn't have any platform-specific recommendations and discovery features, just audio streaming and organizing by tags. I'm also using druid!
It'll be released (hopefully soon) under Apache or MIT.

**Reply by erikas-taroza** [reply]
Hello everyone. I am also trying to implement streaming audio. I have made my own struct that implements `MediaSource`. It works fine if you are reading from start to end. However, I am having lots of trouble with seeking. It seems like the differences between the decoders is the reason this is difficult.

@jpochyla I took some time to look through your code, but I couldn't figure out what exactly was happening. Can you please explain to me how you fixed this problem? Thank you.

@pdeljanov Do you have any ideas on how to implement streaming? Please let us know. Thank you.

**Reply by pdeljanov** [reply]
Hi @erikas-taroza,

Could you explain what the problem with seeking you're having is?

Symphonia just uses the `std::io::Seek` trait so the implementation should be straightforward. You will need to have the entirety of the media on-hand though to support the different seek modes.

However, if your underlying source is unseekable or you don't have all the data, then you can either return `false` in `MediaSource::is_seekable` and `panic!` if `Seek::seek` is called. Or, you can wrap your `std::io::Read`er in a  `symphonia::core::io::ReadOnlySource`.

**Reply by erikas-taroza** [reply]
Hello @pdeljanov 

> Could you explain what the problem with seeking you're having is?

I have been testing with .mp4 and .mp3 files. In the case of mp3 files, when seeking, `std::io::Seek` doesn't actually get called. Instead, it keeps on calling `std::io::Read` to get to the correct bytes for the seek location. With .mp4, `std::io::Seek` is called, but it doesn't actually seek to the correct spot (maybe issue with my implementation). Instead, the sound does a skip and continues as if there wasn't a seek. It does eventually stop reading after the seek, even though there are more bytes available.

> You will need to have the entirety of the media on-hand though to support the different seek modes.

I am streaming by requesting chunks of the audio file with HTTP and the Range header. That means I do not have access to the whole file in the beginning.

> However, if your underlying source is unseekable or you don't have all the data, then you can either return false in MediaSource::is_seekable and panic! if Seek::seek is called.

This would not work for me. I make a new request after the reader is nearly out of bytes, so doing this would make it impossible to seek. When I was planning, I thought I could just fill the buffer with 0s until the seek position, and then make a request at the seek position.

[This is what I have so far.](https://github.com/erikas-taroza/simple_audio/blob/streaming/rust/src/audio/streamable_file.rs)

**Reply by pdeljanov** [reply]
Okay, I understand what you're trying to do now. Basically you want to stream a media file over HTTP.

> I have been testing with .mp4 and .mp3 files. In the case of mp3 files, when seeking, `std::io::Seek` doesn't actually get called. Instead, it keeps on calling `std::io::Read` to get to the correct bytes for the seek location. With .mp4, `std::io::Seek` is called, but it doesn't actually seek to the correct spot (maybe issue with my implementation). Instead, the sound does a skip and continues as if there wasn't a seek. It does eventually stop reading after the seek, even though there are more bytes available.

To clear this up a bit. It's actually the format/media container not the codec that determines how seeking is performed. For example, a MP3 encoded audio track in a MP4 file will seek like an MP4.

MP3 in it's native container (an MPEG Elementary Stream, aka, .mp3 file) does not contain any timing information. So the only way to do an accurate seek is to `read` the file from the beginning and add up the number of samples until we reach the sample we want. This is as inefficient as it sounds! Therefore, you will only see a single `seek` back to the beginning of the file (if the timestamp is in the past), and then a bunch of `read`s. Starting with version v0.5.2, coarse seeking is implemented for the MP3 format by using the average bitrate, file duration, and total duration to guess where to seek to. In this mode, you will only seek one `seek` and then playback resumes.

So depending on the format, the number of reads and seeks will be different.

- MP1/2/3 and ADTS behave as described above.
- MP4 contains tables that tell the reader where to seek to directly, so only a single `seek` is required.
- FLAC contains an optional table of coarse locations where to `seek` to, but then you must `read` until you find the exact location. If the table is not present, then a binary search across the entire file is used to find the coarse location. Expect many `seek`s followed by `read`s.
- WAV only needs to do a single `seek` since the size of each sample is known
- OGG does not contain any seek tables, so a binary search is performed across the entire file to find the coarse location. Expect many `seek`s followed by `read`s.

> > You will need to have the entirety of the media on-hand though to support the different seek modes.
> 
> I am streaming by requesting chunks of the audio file with HTTP and the Range header. That means I do not have access to the whole file in the beginning.

Right. I was a bit imprecise. If you know the total size of the media, and can get ranges of bytes for that file, then a `MediaSource` can be seekable since it's possible to pretend the source is a file.

> > However, if your underlying source is unseekable or you don't have all the data, then you can either return false in MediaSource::is_seekable and panic! if Seek::seek is called.
> 
> This would not work for me. I make a new request after the reader is nearly out of bytes, so doing this would make it impossible to seek. When I was planning, I thought I could just fill the buffer with 0s until the seek position, and then make a request at the seek position.
> 
> [This is what I have so far.](https://github.com/erikas-taroza/simple_audio/blob/streaming/rust/src/audio/streamable_file.rs)

It's quite late where I am, so I can't really dive into code. However, I can dump my thoughts here that should help you implement a solution.

Your media source basically needs to know and keep track of two things: the file size, and the current position in the file.

The file size is easy, a HTTP HEAD request should be able to get you it when you initialize your `MediaSource`.

Implementing `Read` is pretty simple, just do a range request between the current position and current position + the read length requested. Increment current position by the amount of bytes the server actually returns. Issuing individual range requests will be slow though [^1].

Implementing `Seek` is a bit trickier. Basically, given the file size, the current position, and the requested seek position, you need to make sure the seek stays within the bounds of the file, and if it does, update the current position accordingly. No actual HTTP request needs to be made.

- If `SeekFrom::Start`, check if the requested position is > file size, if not, set the current position to the requested position.
- If `SeekFrom::Current`, check if adding the requested relative position to the current position remains within the bounds of the file. If it does, set the current position to the current position + the requested relative position.
- If `SeekFrom::End`, check if adding the requested relative position to the file size remains within the bounds of the file. If it does, set the current position to the current position + requested relative position.

If successful, return the new current position. Otherwise, return an error.

This should effectively emulate reading a local file but over HTTP. 

[^1]: A more advanced solution would be downloading the data on a background thread in one continuous range request and service the `read` requests from that downloaded file. A `seek` would interrupt this download, but after settling the continuous download would resume.

**Reply by erikas-taroza** [reply]
Thank you so much for all this detail! I will take into account what you have said here. I will provide my code as a solution here once I have things working to my liking.

**Reply by erikas-taroza** [reply]
Hello @pdeljanov

I have been working on improving my implementation as well as seeking. So far, I have tested with MP3 and WAV (using coarse seeking) which works well. However, with MP4, it is not as you described above. I can freely seek in the downloaded areas. However, when seeking to a place that is not downloaded, it does as I described earlier. Any ideas?

**Reply by pdeljanov** [reply]
Hi @erikas-taroza,

If you can seek the file with `symphonia-play` then it's extremely unlikely the problem is with Symphonia. Please check it with that first (`-s` option).

Once I know if that works I can provide more thoughts.

**Reply by erikas-taroza** [reply]
Hello @pdeljanov 

If I download the file and play it like so: `cargo run ./test.mp4 -s 100`, it works.
If I do `curl -s https://myurl.com | cargo run - -s 0`, like the README states, it works.
If I do `curl -s https://myurl.com | cargo run - -s 100`, the file does not play.

Before working on streaming, I first downloaded the whole file to RAM before playing it. Seeking worked fine in that case.

**Reply by pdeljanov** [reply]
Okay, that seems reasonable. The failure with the second curl run indicates to me that the MP4 you're using is not optimized for streaming. That is to say that the sample tables (the tables which indicate the offset of each packet, aka "samples" in MP4 jargon) are at the end of the file. This is typical, but it means that since you cannot seek a pipe, the MP4 is unplayable.

However, since you're able to implement `Seek` this is not a problem to be concerned about. You'll probably just notice that when you play this file there will be some initial seeks before playback as all the tables are loaded into memory.

So I looked at your code and I saw that you are working to implement the optimized background thread method. This will definitely be a great solution, but I am concerned about the complexity at this stage.

I feel like you've focused very strongly on how Symphonia accesses a file, but that's going to be a losing battle. My descriptions were very simplified since those access patterns change based on the container format, and how the file was muxed (like the example above). Also, the file access pattern isn't defined or guaranteed by Symphonia, so it's best not to rely on it.

Ultimately, Symphonia depends on the behavior guaranteed by the `Read` and `Seek` traits, and that the correct data is read at the correct position. Therefore, I think to move forward you need to prove the correctness of your implementation of these traits on your source.

You should do this without Symphonia. I would suggest writing standalone tests that exercise the read and seek functions thoroughly. I'd personally do this by generating a large file with random bytes, and then performing reads and seeks on it over both HTTP and via. a `File` directly and comparing what's read. They should match. Make sure to test many back-to-back seeks, seeks to odd offsets, seeks to non-chunk aligned positions, seeks to the start and end of file, reads of odd sizes, etc.

**Reply by erikas-taroza** [reply]
> The failure with the second curl run indicates to me that the MP4 you're using is not optimized for streaming.

That's weird. The files I am using for testing are scraped from YouTube so I would assume they are optimized for streaming.

> You'll probably just notice that when you play this file there will be some initial seeks before playback as all the tables are loaded into memory.

Yes, this is what I noticed.

I will test as you have suggested and see if anything comes up.

Thank you.

**Reply by pdeljanov** [reply]
> > The failure with the second curl run indicates to me that the MP4 you're using is not optimized for streaming.
> 
> That's weird. The files I am using for testing are scraped from YouTube so I would assume they are optimized for streaming.
> 

I believe by default `youtube-dl` and its forks "fixup" (remux) the file as part of the process. If you don't do a fixup, then you actually get a fragmented MP4 which is even more different. Symphonia does support fragmented MP4s though.

**Reply by erikas-taroza** [reply]
Hello @pdeljanov 

I have tested my streaming file vs a normal Rust file. Initially, my file had some issues with requesting chunks beyond the bounds (which the http server did not like). Another thing I noticed was that the normal file allowed seeks to beyond the file size, where my implementation just returned the read position as the seek position. The tests that had normal/in bound numbers did pass as expected though. [Here](https://github.com/erikas-taroza/streaming_test) is my test project if you need to take a look.

I started to test again with Symphonia and I did find one issue was with how I was buffering. I would do an early return with `Ok(buf.len)` which was very incorrect. This caused a seek with a very big offset (>2 billion). I have fixed this issue.

Now when loading the file, it keeps on seeking and reading until the end of the file is reached, essentially downloading the whole file. Only then is playback started. I did some debugging and found [this line](https://github.com/pdeljanov/Symphonia/blob/412f44daab39920beeb81d78b0e4271b263d33e9/symphonia-format-isomp4/src/atoms/mod.rs#L419) as the initiator of these seek calls. Is this the result of the fragmented MP4 that YouTube serves? If it is, then how do I go around fixing it, if it's even possible?

**Reply by erikas-taroza** [reply]
I found a link to a normal MP4 file and I was able to stream it, with seeking, without any issues.

I spent some time with https://www.onlinemp4parser.com/ and ChatGPT to learn how the MP4 format works. I downloaded a fragmented MP4 and also made a copy with ffmpeg that is not fragmented. I found that the STBL atom in the fragmented MP4 is pretty much empty when compared to the normal file. I also noticed that in the TRUN atom of every MOOF atom, the sample size entries corresponded to those in the STSZ atom of the normal file. This maybe explains why I was previously only able to freely seek in downloaded parts.

I am not really sure where to go from here. Maybe course seeking for MP4 would work?

**Reply by pdeljanov** [reply]
If I understand correctly, it sounds like the `MediaSource` works properly now and the only problem is that loading the MP4 is slow?

What you're seeing sounds expected to me. The reader scans the top-level atoms in `try_new` to make sure it's playable. With a fragmented MP4, there are a lot of top-level atoms (mostly MOOF + MDAT pairs). While scanning, the function `ignore_bytes` is used to skip the body of the atoms. Now, `ignore_bytes` will perform a seek if the count is large enough, but if not, it just reads until it has skipped enough. I looked at a YouTube MP4 and the fragments are pretty short (10 seconds) and may fall below the threshold required for a seek. Therefore, it's likely that the file is just being read entirely start to finish.

I think there are few avenues for optimization (from easiest to hardest):

1. I think `ignore_bytes` can be replaced by an explicit `seek` if `is_seekable` is true. However, you will still be doing a lot of seek + small read operations on the first load as the atoms are skipped over.
2. If the MOOV and FTYP atoms were already encountered, it may be reasonable to terminate scanning of the top-level atoms at the first MOOF or MDAT atom. I'd need some documentation (i.e., standards, etc.) or proof that this is a good general optimization that won't break anything. However, this would still defer loading the fragments to the first seek. I think that may be more reasonable though.
3. Use the SIDX atom (segment index) to avoid the scanning that would happen on a seek. This is likely a very tricky thing to implement.

Let's move further discussion over to #175 since this likely isn't the best place for such a discussion.

**Comment by aschey** [comment]
If anyone else stumbles on this discussion looking for a way to add streaming to their project, I made a crate for this that should hopefully be flexible enough for most use cases: https://github.com/aschey/stream-download-rs.
